File: tgrelay/terminator.py
"""Talking to the process that owns Nova's TLS handshake.

The relay's outbound TLS used to be CPython's OpenSSL, whose ClientHello
matches no browser on any desktop — measured, `t13d181100`, a fingerprint that
identifies Nova and nothing else. The handshake now happens in a helper built
from `nova-rs/crates/nova-tls`, which emits the shape of a captured browser and
hands back a plaintext stream over loopback.

The one thing this module exists to get right is the *phase*. `phase.py` decides
which layer to blame by looking at what Python raised — an `ssl.SSLError`, a
`ConnectionResetError`, a Windows error number. None of those can happen once
the socket Python holds is a loopback hop: every failure would arrive as a bare
EOF and collapse into a single verdict. That verdict happens to be the one that
rotates the TLS profile *and* retires the zone's neutral SNI for fifteen
minutes, so a refused connection to a dead egress would quietly cost both.

So the helper reports `reached` and `ended` as numbers, this module copies them
onto the exception, and `phase.ended_from_exception` prefers them over anything
it could infer. Nothing downstream changes.

Disabled unless `NOVA_TLS_TERMINATOR_PORT` is set, and the relay keeps its old
path in that case — this is a lever, not a migration.
"""

import asyncio
import json
import os
from typing import Dict, List, Optional, Tuple

from . import phase


def _env_int(name: str, default: int = 0) -> int:
    try:
        return int(str(os.environ.get(name, "") or default).strip())
    except (TypeError, ValueError):
        return default


def runtime_path() -> str:
    """Where the supervisor publishes the port and token for this run.

    A file rather than environment variables because there are two client
    processes, not one: the relay runs on a thread inside Nova itself, and
    `NovaWFP/proxy/tcp_proxy.py` is a separate interpreter that imports the same
    relay module. A variable exported after that process started would never
    reach it.

    Under `temp/` because both values are regenerated every run — the token is
    not a secret to keep, it is a secret to hold for as long as the helper lives.
    """
    root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    return os.path.join(root, "temp", "tls_terminator.json")


def _runtime() -> Dict[str, object]:
    try:
        with open(runtime_path(), "r", encoding="utf-8") as handle:
            data = json.load(handle)
        return data if isinstance(data, dict) else {}
    except Exception:
        return {}


def terminator_port() -> int:
    """The loopback port the helper listens on, or 0 when it is not in use."""
    from_env = _env_int("NOVA_TLS_TERMINATOR_PORT", 0)
    if from_env > 0:
        return from_env
    try:
        return int(_runtime().get("port") or 0)
    except (TypeError, ValueError):
        return 0


def terminator_token() -> str:
    """The shared secret. Loopback is not a permission boundary on Windows.

    Without this, any process on the machine could use the helper as an open
    TLS proxy wearing a browser's fingerprint, which is a more useful thing to
    steal than an ordinary open proxy.
    """
    token = str(os.environ.get("NOVA_TLS_TERMINATOR_TOKEN", "") or "").strip()
    return token or str(_runtime().get("token") or "").strip()


def active_profile() -> str:
    """Which shape to ask for. One name drives this and the HTTP persona."""
    return str(os.environ.get("NOVA_TLS_PROFILE", "") or "yandex-windows").strip()


def is_enabled() -> bool:
    return terminator_port() > 0 and bool(terminator_token())


def _egress_of(attempt: Dict[str, object]) -> Optional[Dict[str, object]]:
    """Translate one of the relay's upstream attempts into the wire form."""
    kind = str(attempt.get("kind") or "").strip().lower()
    if kind == "direct":
        return {"kind": "direct"}
    host = str(attempt.get("host") or "127.0.0.1").strip()
    port = int(attempt.get("port") or 0)
    if kind == "socks5":
        return {"kind": "socks5", "host": host, "port": port}
    if kind == "http":
        return {"kind": "http", "host": host, "port": port}
    return None


def _tag(exc: BaseException, reached: int, ended: int) -> BaseException:
    """Carry the helper's verdict on the exception, the way transport.py does.

    Annotating rather than wrapping keeps every caller that catches `OSError`
    working untouched.
    """
    try:
        if not hasattr(exc, "nova_reached"):
            exc.nova_reached = int(reached)
        if not hasattr(exc, "nova_ended"):
            exc.nova_ended = int(ended)
    except Exception:
        pass
    return exc


async def _dial_once(
    target_host: str,
    target_port: int,
    sni: str,
    attempt: Dict[str, object],
    connect_timeout: float,
    handshake_timeout: float,
    verify: bool,
):
    egress = _egress_of(attempt)
    if egress is None:
        raise _tag(OSError(f"unsupported egress kind {attempt.get('kind')!r}"),
                   phase.Reached.NOTHING, phase.Ended.CLOSED)

    port = terminator_port()
    token = terminator_token()
    # The helper answers immediately or not at all; it is a local process.
    reader, writer = await asyncio.wait_for(
        asyncio.open_connection("127.0.0.1", port), timeout=5.0
    )
    try:
        greeting = await asyncio.wait_for(reader.readline(), timeout=5.0)
        if not greeting:
            raise _tag(OSError("TLS helper closed without a greeting"),
                       phase.Reached.NOTHING, phase.Ended.CLOSED)

        request = {
            "token": token,
            "target": {"host": str(target_host), "port": int(target_port)},
            "egress": egress,
            "sni": str(sni or target_host),
            "profile": active_profile(),
            "connect_timeout_ms": max(200, int(connect_timeout * 1000)),
            "handshake_timeout_ms": max(200, int(handshake_timeout * 1000)),
            "verify": bool(verify),
        }
        writer.write((json.dumps(request) + "\n").encode("utf-8"))
        await writer.drain()

        # The whole budget, because the helper is doing the connect and the
        # handshake on our behalf and both are inside this wait.
        budget = connect_timeout + handshake_timeout + 2.0
        line = await asyncio.wait_for(reader.readline(), timeout=budget)
        if not line:
            raise _tag(OSError("TLS helper closed before answering"),
                       phase.Reached.NOTHING, phase.Ended.CLOSED)
        status = json.loads(line.decode("utf-8", errors="replace"))
    except BaseException as exc:
        _abort(writer)
        # A failure talking to the helper itself is not evidence about the
        # network, and must not be charged to a domain or an egress.
        raise _tag(exc, phase.Reached.NOTHING, phase.ended_from_exception(exc))

    if not status.get("ok"):
        _abort(writer)
        message = str(status.get("error") or "TLS helper refused the tunnel")
        raise _tag(
            OSError(message),
            int(status.get("reached", phase.Reached.NOTHING)),
            int(status.get("ended", phase.Ended.CLOSED)),
        )
    return reader, writer


def _abort(writer) -> None:
    """Drop a half-built tunnel now, without awaiting.

    Abrupt rather than graceful: the helper reads a loopback close as "the
    relay gave up" and resets its upstream, which is what should happen to a
    tunnel that never became one.
    """
    try:
        transport = getattr(writer, "transport", None)
        if transport is not None and hasattr(transport, "abort"):
            transport.abort()
        else:
            writer.close()
    except Exception:
        pass


async def open_shaped_stream(
    target_host: str,
    target_port: int = 443,
    server_hostname: Optional[str] = None,
    timeout: float = 10.0,
    attempts: Optional[List[Dict[str, object]]] = None,
    verify: bool = False,
) -> Tuple[object, object, str]:
    """Same contract as `transport.open_tls_stream`, one process further out.

    The egress list is walked here rather than inside the helper so that the
    route label the relay logs and learns from stays the label it chose.
    """
    chosen = list(attempts or [])
    if not chosen:
        raise _tag(OSError("no upstream attempts available"),
                   phase.Reached.NOTHING, phase.Ended.CLOSED)

    last_error: Optional[BaseException] = None
    for attempt in chosen:
        label = str(attempt.get("label") or attempt.get("kind") or "unknown").strip() or "unknown"
        try:
            connect_timeout = max(0.2, float(attempt.get("timeout") or timeout))
        except (TypeError, ValueError):
            connect_timeout = float(timeout)
        try:
            reader, writer = await _dial_once(
                target_host,
                int(target_port),
                str(server_hostname or target_host),
                attempt,
                connect_timeout,
                float(timeout),
                verify,
            )
            return reader, writer, label
        except BaseException as exc:
            last_error = exc
            continue

    raise last_error if last_error is not None else _tag(
        OSError("no upstream attempts available"), phase.Reached.NOTHING, phase.Ended.CLOSED
    )

File: nova-rs/crates/nova-tgrelay-net/src/terminator.rs
//! Layer 15: opening a shaped TLS tunnel through the terminator.
//!
//! The port of `tgrelay/terminator.py` — the client half of the loopback
//! protocol whose server is `nova-tls`'s `tls-terminator` binary. This is the
//! last piece of `_connect_websocket_once`, and with it the relay can raise a
//! WSS connection without Python.
//!
//! The wire protocol, taken from the server rather than from the Python client:
//!
//! ```text
//! S->C  {"nova":"tls-terminator/1"}
//! C->S  {"token":"…","target":{"host":"…","port":443},"egress":{…},
//!        "sni":"…","profile":"yandex-windows",
//!        "connect_timeout_ms":1600,"handshake_timeout_ms":8000}
//! S->C  {"ok":true,"reached":4,"ended":0,"alpn":"http/1.1"}
//!       {"ok":false,"reached":1,"ended":3,"error":"connection refused"}
//! ```
//!
//! Then the same socket carries plaintext until either side closes.
//!
//! **The request line must be the last thing written before the reply is read.**
//! The server reads it through a `BufReader` it discards immediately after, so
//! anything pipelined behind it is dropped without a trace. Here that is
//! structural: [`open_shaped_stream`] does not hand the stream back until the
//! reply has arrived, so there is no window in which a caller could write early.
//!
//! **Why the egress list is walked here and not inside the helper.** The route
//! label the relay logs and learns from has to stay the label the relay chose;
//! a helper picking its own would report an egress the health tables never
//! selected.

use nova_probe::Reached;
use nova_tgrelay::egress::{Egress, ProxyProtocol};
use nova_tgrelay::Authority;
use serde::{Deserialize, Serialize};
use std::io;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::TcpStream;
use tokio::time::{timeout, Duration};

/// How long the helper gets to accept a loopback connection and greet.
///
/// It answers immediately or not at all: it is a local process, and anything
/// slower than this is a helper that has died rather than one that is busy.
pub const HELPER_GREETING_TIMEOUT: Duration = Duration::from_secs(5);
/// Added to the connect and handshake budgets when waiting for the reply, since
/// the helper is doing both on our behalf inside that one wait.
pub const REPLY_SLACK: Duration = Duration::from_secs(2);
/// `max(0.2, …)` — no attempt is given less than this.
pub const MIN_ATTEMPT_TIMEOUT: Duration = Duration::from_millis(200);

/// Where the helper is and what it will accept.
#[derive(Debug, Clone)]
pub struct TerminatorConfig {
    pub port: u16,
    /// Minted per run. Without it the helper is an open TLS proxy for every
    /// process on the machine, which is why it refuses to start without one.
    pub token: String,
    /// Which ClientHello shape to emit. `None` leaves the helper on its own
    /// default; naming it per attempt is what lets a learner try shapes without
    /// restarting the helper.
    pub profile: Option<String>,
    /// Whether the helper verifies the peer certificate. The relay's current
    /// behaviour is not to.
    pub verify: bool,
}

#[derive(Debug, Serialize)]
struct WireTarget<'a> {
    host: &'a str,
    port: u16,
}

/// Serialised exactly as the server's `#[serde(tag = "kind", rename_all =
/// "snake_case")]` enum expects: `direct`, `socks5`, `http`.
#[derive(Debug, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
enum WireEgress<'a> {
    Direct,
    Socks5 { host: &'a str, port: u16 },
    Http { host: &'a str, port: u16 },
}

#[derive(Debug, Serialize)]
struct WireRequest<'a> {
    token: &'a str,
    target: WireTarget<'a>,
    egress: WireEgress<'a>,
    sni: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    profile: Option<&'a str>,
    connect_timeout_ms: u64,
    handshake_timeout_ms: u64,
    verify: bool,
}

/// The helper's answer.
#[derive(Debug, Deserialize, PartialEq, Eq)]
pub struct TerminatorReply {
    pub ok: bool,
    #[serde(default)]
    pub reached: u8,
    #[serde(default)]
    pub ended: u8,
    #[serde(default)]
    pub alpn: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
}

/// A tunnel the helper handshook on our behalf.
#[derive(Debug)]
pub struct Shaped {
    pub stream: TcpStream,
    /// The label the *relay* chose, not one the helper invented.
    pub label: String,
    pub alpn: Option<String>,
    /// Bytes the helper sent after its reply, already taken off the socket.
    ///
    /// Empty in every case seen so far — on this path the relay writes the
    /// upgrade request before the far side says anything — but the reply is read
    /// through a `BufReader`, and `into_inner` **drops whatever that buffer still
    /// holds**. Handing back the stream without these would lose them silently,
    /// which is the same over-read this codebase already fixed in the SOCKS
    /// handshake and the CONNECT reply.
    pub leftover: Vec<u8>,
}

#[derive(Debug)]
pub enum TerminatorError {
    NoEgress,
    /// Talking to the helper itself failed.
    ///
    /// **Not evidence about the network**, and must not be charged to a domain
    /// or an egress — the Python tags it `Reached::Nothing` for exactly that
    /// reason. A helper that died takes the whole relay's TLS masking with it
    /// (G20) and says nothing about the route that was being attempted.
    Helper(io::Error),
    /// The helper answered and refused, carrying its own verdict on how far the
    /// attempt got. That verdict *is* about the network.
    Refused {
        reached: u8,
        ended: u8,
        message: String,
        label: String,
    },
}

impl std::fmt::Display for TerminatorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NoEgress => write!(f, "no upstream attempts available"),
            Self::Helper(e) => write!(f, "TLS helper: {e}"),
            Self::Refused { reached, ended, message, label } => {
                write!(f, "TLS helper refused via {label} (reached {reached}, ended {ended}): {message}")
            }
        }
    }
}

impl std::error::Error for TerminatorError {}

impl TerminatorError {
    /// How far the attempt got, in the shared vocabulary.
    pub fn reached(&self) -> Reached {
        match self {
            Self::NoEgress | Self::Helper(_) => Reached::Nothing,
            Self::Refused { reached, .. } => reached_from(*reached),
        }
    }
}

/// The numbers the helper sends are `phase.Reached`'s, verbatim.
fn reached_from(value: u8) -> Reached {
    match value {
        1 => Reached::Resolved,
        2 => Reached::Connected,
        3 => Reached::HelloSent,
        4 => Reached::HandshakeDone,
        5 => Reached::Upgraded,
        6 => Reached::Carrying,
        _ => Reached::Nothing,
    }
}

fn wire_egress(egress: &Egress) -> WireEgress<'_> {
    match egress {
        Egress::Direct { .. } => WireEgress::Direct,
        Egress::Proxy { protocol: ProxyProtocol::Socks5, at, .. } => {
            WireEgress::Socks5 { host: at.host(), port: at.port() }
        }
        Egress::Proxy { protocol: ProxyProtocol::HttpConnect, at, .. } => {
            WireEgress::Http { host: at.host(), port: at.port() }
        }
    }
}

/// The request line, newline included.
///
/// Exposed so a test can assert the bytes: the server on the other end is a
/// separate binary shipped separately, and a field renamed on one side only is
/// a failure that appears in production and nowhere else.
pub fn request_line(
    target: &Authority,
    egress: &Egress,
    sni: &str,
    config: &TerminatorConfig,
    connect_timeout: Duration,
    handshake_timeout: Duration,
) -> String {
    let request = WireRequest {
        token: &config.token,
        target: WireTarget { host: target.host(), port: target.port() },
        egress: wire_egress(egress),
        sni,
        profile: config.profile.as_deref(),
        connect_timeout_ms: connect_timeout.max(MIN_ATTEMPT_TIMEOUT).as_millis() as u64,
        handshake_timeout_ms: handshake_timeout.max(MIN_ATTEMPT_TIMEOUT).as_millis() as u64,
        verify: config.verify,
    };
    let mut line = serde_json::to_string(&request).expect("a fixed struct of strings and numbers");
    line.push('\n');
    line
}

/// One egress, one loopback connection to the helper.
async fn dial_once(
    target: &Authority,
    egress: &Egress,
    sni: &str,
    config: &TerminatorConfig,
    connect_timeout: Duration,
    handshake_timeout: Duration,
) -> Result<Shaped, TerminatorError> {
    let helper = format!("127.0.0.1:{}", config.port);
    let stream = timeout(HELPER_GREETING_TIMEOUT, TcpStream::connect(&helper))
        .await
        .map_err(|_| TerminatorError::Helper(io::Error::from(io::ErrorKind::TimedOut)))?
        .map_err(TerminatorError::Helper)?;
    let _ = stream.set_nodelay(true);

    let mut reader = BufReader::new(stream);
    let mut greeting = String::new();
    let read = timeout(HELPER_GREETING_TIMEOUT, reader.read_line(&mut greeting))
        .await
        .map_err(|_| TerminatorError::Helper(io::Error::from(io::ErrorKind::TimedOut)))?
        .map_err(TerminatorError::Helper)?;
    if read == 0 {
        return Err(TerminatorError::Helper(io::Error::other("TLS helper closed without a greeting")));
    }

    let line = request_line(target, egress, sni, config, connect_timeout, handshake_timeout);
    reader.get_mut().write_all(line.as_bytes()).await.map_err(TerminatorError::Helper)?;
    reader.get_mut().flush().await.map_err(TerminatorError::Helper)?;

    // The helper is doing the connect *and* the handshake inside this one wait.
    let budget = connect_timeout + handshake_timeout + REPLY_SLACK;
    let mut answer = String::new();
    let read = timeout(budget, reader.read_line(&mut answer))
        .await
        .map_err(|_| TerminatorError::Helper(io::Error::from(io::ErrorKind::TimedOut)))?
        .map_err(TerminatorError::Helper)?;
    if read == 0 {
        return Err(TerminatorError::Helper(io::Error::other("TLS helper closed before answering")));
    }
    let reply: TerminatorReply = serde_json::from_str(answer.trim_end())
        .map_err(|e| TerminatorError::Helper(io::Error::other(format!("bad reply: {e}"))))?;

    if !reply.ok {
        // Dropped abruptly rather than closed politely: the helper reads a
        // loopback close as "the relay gave up" and resets its upstream, which
        // is what should happen to a tunnel that never became one.
        return Err(TerminatorError::Refused {
            reached: reply.reached,
            ended: reply.ended,
            message: reply.error.unwrap_or_else(|| "TLS helper refused the tunnel".to_string()),
            label: egress.label().to_string(),
        });
    }

    // Taken *before* `into_inner`, which throws the buffer away.
    let leftover = reader.buffer().to_vec();
    let stream = reader.into_inner();
    Ok(Shaped { stream, label: egress.label().to_string(), alpn: reply.alpn, leftover })
}

/// Same contract as the plain dialler, one process further out.
///
/// Walks the egresses in the order given and returns the first tunnel the helper
/// manages to raise.
pub async fn open_shaped_stream(
    target: &Authority,
    sni: &str,
    egresses: &[Egress],
    config: &TerminatorConfig,
    default_timeout: Duration,
) -> Result<Shaped, TerminatorError> {
    if egresses.is_empty() {
        return Err(TerminatorError::NoEgress);
    }
    let mut last = None;
    for egress in egresses {
        let connect_timeout =
            egress.timeout().filter(|t| !t.is_zero()).unwrap_or(default_timeout).max(MIN_ATTEMPT_TIMEOUT);
        match dial_once(target, egress, sni, config, connect_timeout, default_timeout).await {
            Ok(shaped) => return Ok(shaped),
            Err(e) => last = Some(e),
        }
    }
    Err(last.unwrap_or(TerminatorError::NoEgress))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn config() -> TerminatorConfig {
        TerminatorConfig {
            port: 1374,
            token: "s3cret".to_string(),
            profile: Some("yandex-windows".to_string()),
            verify: false,
        }
    }

    fn target() -> Authority {
        Authority::new("kws2.nova-app.eu", 443).expect("authority")
    }

    fn warp() -> Egress {
        Egress::proxy(ProxyProtocol::Socks5, "warp-socks", Authority::new("127.0.0.1", 1370).unwrap())
    }

    #[test]
    fn the_request_is_the_shape_the_helper_deserialises() {
        // The server is a separately shipped binary; a field renamed on one side
        // only fails in production and nowhere else.
        let line = request_line(
            &target(),
            &warp(),
            "www.nova-app.eu",
            &config(),
            Duration::from_millis(1600),
            Duration::from_millis(8000),
        );
        assert!(line.ends_with('\n'), "one line, newline-terminated");
        let parsed: serde_json::Value = serde_json::from_str(line.trim_end()).expect("valid json");
        assert_eq!(parsed["token"], "s3cret");
        assert_eq!(parsed["target"]["host"], "kws2.nova-app.eu");
        assert_eq!(parsed["target"]["port"], 443);
        assert_eq!(parsed["egress"]["kind"], "socks5");
        assert_eq!(parsed["egress"]["host"], "127.0.0.1");
        assert_eq!(parsed["egress"]["port"], 1370);
        assert_eq!(parsed["sni"], "www.nova-app.eu");
        assert_eq!(parsed["profile"], "yandex-windows");
        assert_eq!(parsed["connect_timeout_ms"], 1600);
        assert_eq!(parsed["handshake_timeout_ms"], 8000);
        assert_eq!(parsed["verify"], false);
    }

    #[test]
    fn each_egress_kind_uses_the_name_the_server_enum_renames_to() {
        let cases = [
            (Egress::direct("direct"), "direct"),
            (warp(), "socks5"),
            (
                Egress::proxy(ProxyProtocol::HttpConnect, "opera-http", Authority::new("127.0.0.1", 1371).unwrap()),
                "http",
            ),
        ];
        for (egress, kind) in cases {
            let line =
                request_line(&target(), &egress, "x", &config(), Duration::from_secs(1), Duration::from_secs(1));
            let parsed: serde_json::Value = serde_json::from_str(line.trim_end()).expect("json");
            assert_eq!(parsed["egress"]["kind"], kind);
            if kind == "direct" {
                assert!(parsed["egress"].get("host").is_none(), "direct carries no address");
            }
        }
    }

    #[test]
    fn a_missing_profile_is_omitted_rather_than_sent_as_null() {
        // The server's field is `Option<String>` with `#[serde(default)]`; a
        // literal `null` would deserialise the same way, but omitting it is what
        // "leave the helper on its own default" should look like on the wire.
        let mut config = config();
        config.profile = None;
        let line = request_line(&target(), &warp(), "x", &config, Duration::from_secs(1), Duration::from_secs(1));
        assert!(!line.contains("profile"), "{line}");
    }

    #[test]
    fn no_attempt_is_given_less_than_the_floor() {
        let line =
            request_line(&target(), &warp(), "x", &config(), Duration::from_millis(1), Duration::from_millis(0));
        let parsed: serde_json::Value = serde_json::from_str(line.trim_end()).expect("json");
        assert_eq!(parsed["connect_timeout_ms"], 200);
        assert_eq!(parsed["handshake_timeout_ms"], 200);
    }

    #[test]
    fn the_replies_the_server_actually_sends_all_parse() {
        let ok: TerminatorReply =
            serde_json::from_str(r#"{"ok":true,"reached":4,"ended":0,"alpn":"http/1.1"}"#).expect("ok");
        assert!(ok.ok);
        assert_eq!(ok.alpn.as_deref(), Some("http/1.1"));
        assert_eq!(reached_from(ok.reached), Reached::HandshakeDone);

        let refused: TerminatorReply =
            serde_json::from_str(r#"{"ok":false,"reached":1,"ended":3,"error":"connection refused"}"#)
                .expect("refused");
        assert!(!refused.ok);
        assert_eq!(reached_from(refused.reached), Reached::Resolved);

        // The terse one the server sends for a bad token, written out by hand
        // there rather than serialised.
        let unauthorised: TerminatorReply =
            serde_json::from_str(r#"{"ok":false,"reached":0,"ended":4,"error":"unauthorised"}"#)
                .expect("unauthorised");
        assert_eq!(unauthorised.error.as_deref(), Some("unauthorised"));

        // And a success with no ALPN, which the server sends as a literal null.
        let no_alpn: TerminatorReply =
            serde_json::from_str(r#"{"ok":true,"reached":4,"ended":0,"alpn":null}"#).expect("no alpn");
        assert_eq!(no_alpn.alpn, None);
    }

    #[test]
    fn a_helper_failure_is_never_charged_to_the_network() {
        // The distinction the whole error type exists for: a dead helper takes
        // the relay's TLS masking with it (G20) and says nothing whatever about
        // the route that was being attempted.
        let helper = TerminatorError::Helper(io::Error::from(io::ErrorKind::ConnectionRefused));
        assert_eq!(helper.reached(), Reached::Nothing);

        let refused = TerminatorError::Refused {
            reached: 3,
            ended: 2,
            message: "reset".to_string(),
            label: "warp-socks".to_string(),
        };
        assert_eq!(refused.reached(), Reached::HelloSent, "the helper's verdict is the network's");
    }

    #[test]
    fn an_unknown_reached_number_degrades_to_nothing() {
        // A future helper could grow a value this build has never heard of.
        // Reading it as `Carrying` would credit a route with a success it never
        // had; `Nothing` concludes the least.
        assert_eq!(reached_from(99), Reached::Nothing);
        assert_eq!(reached_from(0), Reached::Nothing);
    }
}
